### PR TITLE
Rewrite libcore's UTF-8 validation for performance

### PR DESCRIPTION
**Pull request by thomcc**
This optimizes the `core::str::from_utf8` function significantly (initial measurements indicates that it's often 1.5x faster, especially for non-ASCII, where it can be up to 3x faster).

It does this mostly by leveraging the [shift-based DFA](https://gist.github.com/pervognsen/218ea17743e1442e59bb60d29b1aa725) technique (a recent obsession), but also it adding SIMD to the ASCII fast path (and really just completely rewrites and restructures how the validation is done).

For prior art: shift-DFAs are now used for UTF-8 validation in [PostgreSQL](https://github.com/postgres/postgres/blob/aa6954104644334c53838f181053b9f7aa13f58c/src/common/wchar.c#L1753), and seems to be in progress or under consideration for use in [Julia](https://github.com/JuliaLang/julia/pull/47880) and perhaps [Go](https://github.com/golang/go/issues/47120). Of these, PG's impl is the most similar to this one, at least at a high level[^pg].

[^pg]: The main similarity is that PG also uses 32bit rows in the transition table, and has a special case for ASCII (even if the way we special case it is very different). Beyond that, the impls are really totally different (theirs has fewer optimizations, doesn't need tracking any info for error positions, uses a different UTF8 automaton, and is just completely different code).

---

This PR is not quite ready for review. I'm mostly getting this up now so I can check some perf runs and such. Assuming they (or further benchmarks) don't reveal issues, I believe the approach is basically complete, but the PR needs the following work to be ready for review I intend to:

- [ ] Add more comments/docs, it's totally impossible to follow at the moment.
- [ ] Rewrite the code for the table so that it's comprehensible instead of a total black box, and include a link to a generator the constants use.
- [ ] Rewrite the ASCII scan loop (which is pretty clownshoes at the moment, and probably should still handle alignment after the first load).
- [ ] Ensure the tests cover all the optimized cases (both short/long strings).
- [ ] Integrate some of my benchmarks[^bench].

[^bench]: The benchmarks I have locally have too much PII to be published as-is, since I derived them from real use including browser history (among other things), but that's a temporary situation. Some preliminary benchmarks based on the `simdutf8` corpus are [here](https://gist.github.com/thomcc/1d03d323aecfed570c77c060e2e1cb63). It demonstrates a speedup in basically every case (across all string sizes and character compositions), although the real-world improvement seems to be even higher (the impl this replaces has branch misprediction issues for non-ASCII, which are not reflected in these benchmarks). 

I'm deliberately leaving anything that touches other functions as follow-up work that I'll do after this lands. That includes improving `String::from_utf8_lossy` or sharing any logic with `is_ascii`

(Note: some stuff came up in $dayjob so it may be a week or two before I finish all this, just wanted to get this up in the meantime so that it's not really in the back of my mind as much)

### Potential Drawbacks

- Weird algo involving a magic table. I will document this way better so that this drawback goes away.
- Slightly slower in the invalid UTF-8 path, as we may re-validate up to 16 bytes. I'll probably fix this for truncation, but even so it is pretty minimal. It's very likely that if the error isn't right at the start of the string, this is version is still much faster.
- Probably worse on some machines, including ones slow (slower than doing scalar operations) unaligned loads, and slow dynamic right shifts (slower than branching).
- Additional 1kb table, although I'm very careful not to touch it for the pure-ASCII path (doing this without eliminating the benefit of the shift-DFA was hard).
- Need `const_eval_select`. This is both because LLVM was doing worse on the `while ...` version of some of the loops, and because `core::simd` isn't const-compatible (trait usage). The function we call would still exist either way though.

### Why not `simdutf8`'s algorithm?

Short version:
- Can't use algos requiring new instructions in libcore for various reasons. Not worth it to maintain just for `-Zbuild-std` users (okay, and `aarch64` users).
- This impl is pretty small, mostly portable (a couple special cases in the ASCII handling, but nothing complex or conditionally compiled), and does not require much target-specific complexity.
- On the bright side, we're faster for strings up to around 40-120 bytes anyway. At least on my machines.

Long version: <https://gist.github.com/thomcc/f153a122f680023f937f2c912978b8e6>.

### CODE REVIEW

**Comment by thomcc** [comment]
Err,
r? @ghost

File: library/core/src/str/validations/utf8_dfa.rs [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+        }
+        if unlikely(state == ERR) {
+            return Err(utf8_find_error(inp, pos, last_state != END));
+        }
+    }
+    // Did we leave the optimized loop in the middle of a UTF-8 sequence?
+    let was_mid_char = state != END;
+    debug_assert!(state != ERR);
+    if !tail.is_empty() {
+        // Check and early return if the last CHUNK_LEN bytes were all ASCII. The
```
Comment (thomcc) [inline library/core/src/str/validations/utf8_dfa.rs:163]: (note to self: reword this comment, since it seems like I got distracted halfway through it)

File: library/core/src/str/validations/utf8_dfa.rs [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+    }
+}
+
+#[inline]
+fn is_ascii_small(s: &[u8]) -> bool {
+    // LLVM seems to get pretty aggressive if we use a loop here, even if we
+    // check the length first, which can end up having some pretty disasterous
+    // impacts on performance, seemingly due to inlining(?). In any case.
+    //
+    // Note that doing this for many sizes
```
Comment (thomcc) [inline library/core/src/str/validations/utf8_dfa.rs:282]: Forgot to finish this comment. Was going to mention that this seems to still beat the naïve version even if the branch on length is hard to predict, but that stops holding if you add more length conditions.

File: library/core/src/str/validations/utf8_dfa.rs [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        if state == ERR { (inp.len() - tail.len(), was_mid_char) } else { (inp.len(), true) };
+    Err(utf8_find_error(inp, index, backup))
+}
+
+#[inline]
+fn backup_not_yet_invalid(inp: &[u8], mut pos: usize) -> usize {
+    debug_assert!(!inp.is_empty() && inp.get(..pos).is_some());
+    while pos != 0 {
+        pos -= 1;
+        let is_cont = (inp[pos] & 0b1100_0000) == 0b1000_0000;
```
Comment (the8472) [inline library/core/src/str/validations/utf8_dfa.rs:199]: Well, this isn't hot code so probably not significant, but the `as i8 >= -64` construction from the old impl should take 1 less instruction since it can use status flags for signed operations for that comparison.

**Comment by Lokathor** [comment]
8kb can be a pretty big in an embedded context. Can we keep the older, small no-table function available, perhaps under some other name or something?

**Comment by the8472** [comment]
> 8kb can be a pretty big in an embedded context. 

We have a bunch of places in the standard library where that's a concern. Ideally we'd have a some umbrella cfg to make that tradeoff.

**Comment by thomcc** [comment]
> 8kb can be a pretty big in an embedded context. Can we keep the older, small no-table function available, perhaps under some other name or something?

Genuinely not sure why I thought it was 8kb, the table is a `[u32; 256]`, so it's 1kb. Also the current version has a 256b table which might be able to be removed after all is said and done.

File: library/core/src/str/validations/utf8_dfa.rs (hunk 3 above)
Comment (thomcc) [reply library/core/src/str/validations/utf8_dfa.rs:199]: Yeah good point. I think there's a function in the parent module too.

**Comment by thomcc** [comment]
@ndinsmore thanks a lot. I think the adjustment you describe to the DFA isn't directly useful, but it gives me a really good idea for a similar change.

> Also important in the julia port I changed the order of the ops so that the state returned is always "clean"

Changing to `state = (row >> state) & $mask` should result in the same code on most architectures, and seems to https://godbolt.org/z/ha6v6EEoq.

It does have a downside in Rust though — it would lead to additional branches in the code when overflow checks were enabled. Avoiding that would require what we're doing here (or using `wrapping_shr` so I'm not sure there's a point). Regarding having a "clean" state this way, the way the code is structured, this isn't an issue. The code drives the DFA is localized to dedicated functions, and always returns "clean" states. The code outside of that does not have to mask to use comparisons.

> The code is a bit faster, and LLVM vectorizes it just as well

I'm surprised you saw a perf difference, but this kind of thing can be very fiddly (I don't see one, but my states are 32 bits, and I'm using Rust not Julia — hard to say). That said, I don't really expect this is the kind of code that LLVM can vectorize no matter how you write it.

**Comment by ndinsmore** [comment]
@thomcc how did you manage to fit the encoding in 32 bits? You need at least 9 states by my count and I can't figure out how you packed them in.

**Comment by pervognsen** [comment]
> @thomcc how did you manage to fit the encoding in 32 bits? You need at least 9 states by my count and I can't figure out how you packed them in.

While 32-bit rows aren't sufficient to represent all 9-state DFAs, some 9-state DFAs are still representable. Soon after the original article @dougallj used an SMT solver to find a 32-bit-compatible encoding of the DFA transitions for UTF-8 validation. That's what Postgres uses and @thomcc appears similar: https://github.com/postgres/postgres/blob/aa6954104644334c53838f181053b9f7aa13f58c/src/common/wchar.c#L1791

**Comment by Sp00ph** [comment]
There don't seem to be that many resources available online on DFA based UTF-8 validation, the best one I could find was this: https://bjoern.hoehrmann.de/utf-8/decoder/dfa/
The DFAs in that article only have 8 states. Why are we using one with 9 states? Also, how can we ever cram 9 states into 32 bits with the shift based approach? The bits per transition must be constant across all states, so at most 3 bits per state for 9 states, which would only leave 8 of those states addressable, right?

**Comment by Sp00ph** [comment]
Alright I understand now how it's possible to use 9 states in a 32 bit DFA. The states aren't numbered 0-8, and instead use non-contiguous (I think?) values in the range 0-31. That also explains why you'd need an SMT solver to come up with the transitions and state numbers. I still don't see why we're using a DFA with 9 states instead of one with 8 though.

**Comment by Sp00ph** [comment]
Oh right, I should've read the article more carefully. The diagrams in there are omitting the error state.

**Comment by thomcc** [comment]
@Sp00ph Sorry, I'll be writing some more documentation soon.

**Comment by JohnCSimon** [comment]
@thomcc Ping from triage:  Can you post your status on this PR?

**Comment by thomcc** [comment]
I've been busy with various issues, but I'll try to get back to this sooner rather than later so that there can be a comparisons between the different approaches for this vs what's used in #111367 (perhaps taking the best of both could be done, although I'm not sure...)

**Comment by JohnCSimon** [comment]
@thomcc 

ping from triage - can you post your status on this PR? This PR has not received an update in a few months. Thank you!

**Comment by oskgo** [comment]
@thomcc

Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this.
Note: if you are going to continue please open the PR BEFORE you push to it, else you won't be able to reopen - this is a quirk of github.
Thanks for your contribution.

@rustbot label: +S-inactive
